### ISSUE TITLE
RES-3781: llama.cpp in Resilient - Portable LLM Inference Engine

### ISSUE THREAD

**Issue by EricSpencer00** [EricSpencer00/Resilient#3781]
## Vision
Build a complete LLM inference engine in Resilient that works across CPU architectures (x86-64, ARM64, RISC-V). Start with TinyLLaMA 1.1B, expand to larger models.

## Architecture
```
resilient/examples/llm/
├── llama_inference.rz       (public API)
├── quantized_ops.rz         (i8 matmul, softmax, rope)
├── gguf_loader.rz           (model weight loading)
├── sampling.rz              (top-k, top-p)
└── backends/
    ├── x86_simd.rz          (AVX2/AVX-512 intrinsics)
    ├── arm_simd.rz          (NEON intrinsics)
    └── fallback.rz          (pure Rust/C backend)

tests/
├── llama_vs_numpy.rz        (validation against reference)
├── inference_correctness.rz (numerical stability)
└── benchmark.rz             (throughput/latency targets)
```

### Phase A: Quantized Ops (weeks 1-2, ~1200 LOC)
- [ ] i8 matmul (core inference loop)
- [ ] RoPE (rotary positional embeddings)
- [ ] Softmax + exp with fixed precision
- [ ] Validation vs NumPy reference
- **Platform**: CPU fallback only

### Phase B: Model Format (weeks 3-4, ~600 LOC)
- [ ] Parse GGUF format (llama.cpp's standard)
- [ ] Load weights into memory
- [ ] Quantization support (Q4, Q5, Q8)

### Phase C: Sampling + Generation (week 5, ~300 LOC)
- [ ] Top-k sampling
- [ ] Top-p (nucleus) sampling
- [ ] Temperature control
- [ ] Stop tokens

### Phase D: SIMD Portability (weeks 6-8, ~3000 LOC)
- [ ] x86-64 backend (AVX2/AVX-512)
- [ ] ARM64 backend (NEON/SVE)
- [ ] Runtime CPU feature detection
- [ ] Fallback graceful degradation

### Phase E: Example Models + Docs (week 9, ~400 LOC)
- [ ] TinyLLaMA 1.1B example
- [ ] Loading + inference walkthrough
- [ ] Performance tuning guide
- [ ] Porting to new architectures

## Target Models
1. **TinyLLaMA 1.1B** (start here, ~500MB GGUF)
2. **Phi 2.7B** (next milestone)
3. **Llama 2 7B** (longer term)

## Technical Challenges
- [ ] Fixed-size arrays vs dynamic allocation (RES-2606 borrowing)
- [ ] SIMD intrinsics coverage (may need FFI for some ops)
- [ ] Memory layout for quantized weights (alignment/padding)
- [ ] Numerical stability in i8 arithmetic

## Success Metrics
- [ ] Can run TinyLLaMA inference in <2s/token on 4-core CPU
- [ ] Works on x86-64, ARM64, RISC-V (with SIMD)
- [ ] Matches llama.cpp output to ±0.01 logit difference
- [ ] Compiles without unsafe blocks (or minimal, justified unsafe)

## Example Usage
```resilient
let model = load_gguf("tinyllama-1.1b-q4.gguf");
let prompt = "What is 2 + 2?";
let output = generate(model, prompt, max_tokens: 32);
println(output);
// → "What is 2 + 2? 2 + 2 = 4. The answer is 4."
```

## Related
- RES-2606: Borrow analysis (for memory patterns)
- RES-3201: Typestate validation (for model state)

## First Step: Prototype i8 matmul in standalone example, validate vs NumPy

**Comment by EricSpencer00** [comment]
Claiming for agent work. This is multi-month as written, so I’ll land a first vertical slice rather than a paper-only change: a small, testable quantized-op example/corpus foundation that future GGUF and backend work can build on.
